File: src/services/audit_service/presentation.rs
use std::collections::BTreeMap;

use serde_json::Value;

use super::models::{AuditPresentation, AuditPresentationMessage};
use crate::types::{AuditAction, AuditEntityType};

pub fn build_audit_presentation(
    action: AuditAction,
    entity_type: AuditEntityType,
    entity_id: Option<i64>,
    entity_name: Option<&str>,
    details: Option<&str>,
) -> Option<AuditPresentation> {
    let parsed_details = details.and_then(parse_details);
    let summary = Some(summary_message(
        action,
        entity_name,
        parsed_details.as_ref(),
    ));
    let target = match action {
        AuditAction::ServerStart | AuditAction::ServerShutdown => Some(server_target()),
        _ => target_message(entity_type, entity_id, entity_name),
    };
    let detail = detail_message(action, parsed_details.as_ref());

    Some(AuditPresentation {
        summary,
        target,
        detail,
    })
}

fn parse_details(raw: &str) -> Option<Value> {
    serde_json::from_str(raw).ok()
}

fn summary_message(
    action: AuditAction,
    entity_name: Option<&str>,
    details: Option<&Value>,
) -> AuditPresentationMessage {
    let mut params = BTreeMap::new();
    if let Some(name) = entity_name {
        params.insert("name".to_string(), Value::String(name.to_string()));
    }

    match action {
        AuditAction::ConfigUpdate
        | AuditAction::ConfigActionExecute
        | AuditAction::AdminDeleteConfig => {
            if let Some(name) = entity_name {
                params.insert("key".to_string(), Value::String(name.to_string()));
            }
            copy_string_param(details, &mut params, "action");
        }
        AuditAction::AdminCreateExternalAuthProvider
        | AuditAction::AdminUpdateExternalAuthProvider
        | AuditAction::AdminDeleteExternalAuthProvider
        | AuditAction::AdminTestExternalAuthProvider
        | AuditAction::ExternalAuthProviderCreate
        | AuditAction::ExternalAuthProviderUpdate
        | AuditAction::ExternalAuthProviderDelete => {
            copy_string_param(details, &mut params, "key");
            copy_string_param(details, &mut params, "slug");
        }
        AuditAction::MinecraftProfileCreate
        | AuditAction::MinecraftProfileRename
        | AuditAction::MinecraftProfileDelete
        | AuditAction::MinecraftTextureUpload
        | AuditAction::MinecraftTextureBind
        | AuditAction::MinecraftTextureDelete
        | AuditAction::MinecraftTextureLibrarySubmit
        | AuditAction::MinecraftTextureLibraryWithdraw
        | AuditAction::MinecraftTextureLibraryApprove
        | AuditAction::MinecraftTextureLibraryReject
        | AuditAction::MinecraftTextureLibraryUnpublish
        | AuditAction::MinecraftTextureReportCreate
        | AuditAction::MinecraftTextureReportAccept
        | AuditAction::MinecraftTextureReportReject
        | AuditAction::YggdrasilAuthenticate
        | AuditAction::YggdrasilRefreshToken
        | AuditAction::YggdrasilInvalidateToken
        | AuditAction::YggdrasilSignout
        | AuditAction::YggdrasilJoinServer => {
            copy_string_param(details, &mut params, "profile_name");
            copy_string_param(details, &mut params, "profile_uuid");
            copy_string_param(details, &mut params, "old_profile_name");
            copy_string_param(details, &mut params, "new_profile_name");
            copy_string_param(details, &mut params, "texture_type");
            copy_string_param(details, &mut params, "texture_hash");
            copy_string_param(details, &mut params, "library_status");
            copy_string_param(details, &mut params, "reason");
            copy_string_param(details, &mut params, "report_status");
            copy_string_param(details, &mut params, "selected_profile_name");
            copy_string_param(details, &mut params, "selected_profile_uuid");
        }
        _ => {}
    }

    AuditPresentationMessage {
        code: action.as_str().to_string(),
        params,
    }
}

fn target_message(
    entity_type: AuditEntityType,
    entity_id: Option<i64>,
    entity_name: Option<&str>,
) -> Option<AuditPresentationMessage> {
    if entity_id.is_none() && entity_name.is_none() {
        return None;
    }

    let mut params = BTreeMap::new();
    if let Some(id) = entity_id {
        params.insert("id".to_string(), Value::Number(id.into()));
    }
    if let Some(name) = entity_name {
        params.insert("name".to_string(), Value::String(name.to_string()));
    }

    Some(AuditPresentationMessage {
        code: entity_type.as_str().to_string(),
        params,
    })
}

fn server_target() -> AuditPresentationMessage {
    AuditPresentationMessage {
        code: "server".to_string(),
        params: BTreeMap::new(),
    }
}

fn detail_message(
    action: AuditAction,
    details: Option<&Value>,
) -> Option<AuditPresentationMessage> {
    let details = details?;
    let mut params = BTreeMap::new();

    match action {
        AuditAction::ConfigUpdate => {
            copy_param(details, &mut params, "value");
            copy_param(details, &mut params, "visibility");
            copy_param(details, &mut params, "prior_visibility");
            Some(message("config_value_updated", params))
        }
        AuditAction::ConfigActionExecute => {
            copy_param(details, &mut params, "action");
            copy_param(details, &mut params, "target_email");
            Some(message("config_action_executed", params))
        }
        AuditAction::UserLogin => {
            copy_param(details, &mut params, "identifier");
            Some(message("user_login_identifier", params))
        }
        AuditAction::AdminCleanupTasks => {
            copy_param(details, &mut params, "removed");
            copy_param(details, &mut params, "finished_before");
            copy_param(details, &mut params, "kind");
            copy_param(details, &mut params, "status");
            Some(message("tasks_cleanup_finished", params))
        }
        AuditAction::TaskRetry => {
            copy_param(details, &mut params, "kind");
            copy_param(details, &mut params, "previous_attempt_count");
            Some(message("task_retry_scheduled", params))
        }
        AuditAction::MailSend => {
            copy_param(details, &mut params, "to_address");
            copy_param(details, &mut params, "template_code");
            copy_param(details, &mut params, "outbox_id");
            Some(message("mail_sent", params))
        }
        AuditAction::MailDeliveryFailed => {
            copy_param(details, &mut params, "to_address");
            copy_param(details, &mut params, "template_code");
            copy_param(details, &mut params, "outbox_id");
            copy_param(details, &mut params, "attempt_count");
            copy_param(details, &mut params, "error");
            Some(message("mail_delivery_failed", params))
        }
        AuditAction::AdminCreateExternalAuthProvider
        | AuditAction::AdminUpdateExternalAuthProvider
        | AuditAction::AdminDeleteExternalAuthProvider
        | AuditAction::ExternalAuthProviderCreate
        | AuditAction::ExternalAuthProviderUpdate
        | AuditAction::ExternalAuthProviderDelete => {
            copy_param(details, &mut params, "key");
            copy_param(details, &mut params, "slug");
            copy_param(details, &mut params, "kind");
            copy_param(details, &mut params, "issuer_url");
            copy_param(details, &mut params, "enabled");
            Some(message("external_auth_provider_changed", params))
        }
        AuditAction::AdminTestExternalAuthProvider => {
            copy_param(details, &mut params, "provider");
            copy_param(details, &mut params, "key");
            copy_param(details, &mut params, "success");
            copy_param(details, &mut params, "slug");
            copy_param(details, &mut params, "kind");
            copy_param(details, &mut params, "issuer_url");
            copy_param(details, &mut params, "enabled");
            Some(message("external_auth_provider_tested", params))
        }
        AuditAction::MinecraftProfileCreate => {
            copy_param(details, &mut params, "profile_uuid");
            copy_param(details, &mut params, "profile_name");
            Some(message("minecraft_profile_created", params))
        }
        AuditAction::MinecraftProfileRename => {
            copy_param(details, &mut params, "profile_uuid");
            copy_param(details, &mut params, "old_profile_name");
            copy_param(details, &mut params, "new_profile_name");
            copy_param(details, &mut params, "temporarily_invalidated_token_count");
            Some(message("minecraft_profile_renamed", params))
        }
        AuditAction::MinecraftProfileDelete => {
            copy_param(details, &mut params, "profile_uuid");
            copy_param(details, &mut params, "profile_name");
            copy_param(details, &mut params, "deleted_texture_count");
            copy_param(details, &mut params, "revoked_token_count");
            Some(message("minecraft_profile_deleted", params))
        }
        AuditAction::MinecraftTextureUpload => {
            copy_param(details, &mut params, "profile_uuid");
            copy_param(details, &mut params, "profile_name");
            copy_param(details, &mut params, "texture_type");
            copy_param(details, &mut params, "texture_hash");
            copy_param(details, &mut params, "texture_model");
            copy_param(details, &mut params, "width");
            copy_param(details, &mut params, "height");
            copy_param(details, &mut params, "file_size");
            Some(message("minecraft_texture_uploaded", params))
        }
        AuditAction::MinecraftTextureBind => {
            copy_param(details, &mut params, "profile_uuid");
            copy_param(details, &mut params, "profile_name");
            copy_param(details, &mut params, "texture_type");
            copy_param(details, &mut params, "texture_hash");
            copy_param(details, &mut params, "texture_model");
            copy_param(details, &mut params, "width");
            copy_param(details, &mut params, "height");
            copy_param(details, &mut params, "file_size");
            Some(message("minecraft_texture_bound", params))
        }
        AuditAction::MinecraftTextureDelete => {
            copy_param(details, &mut params, "profile_uuid");
            copy_param(details, &mut params, "profile_name");
            copy_param(details, &mut params, "texture_type");
            copy_param(details, &mut params, "texture_hash");
            Some(message("minecraft_texture_deleted", params))
        }
        AuditAction::MinecraftTextureLibrarySubmit
        | AuditAction::MinecraftTextureLibraryWithdraw
        | AuditAction::MinecraftTextureLibraryApprove
        | AuditAction::MinecraftTextureLibraryReject
        | AuditAction::MinecraftTextureLibraryUnpublish => {
            copy_param(details, &mut params, "texture_type");
            copy_param(details, &mut params, "texture_hash");
            copy_param(details, &mut params, "texture_model");
            copy_param(details, &mut params, "library_status");
            copy_param(details, &mut params, "review_note");
            Some(message("minecraft_texture_library_review_changed", params))
        }
        AuditAction::MinecraftTextureReportCreate
        | AuditAction::MinecraftTextureReportAccept
        | AuditAction::MinecraftTextureReportReject => {
            copy_param(details, &mut params, "texture_id");
            copy_param(details, &mut params, "report_id");
            copy_param(details, &mut params, "reason");
            copy_param(details, &mut params, "report_status");
            copy_param(details, &mut params, "library_status");
            Some(message("minecraft_texture_report_changed", params))
        }
        AuditAction::YggdrasilAuthenticate => {
            copy_param(details, &mut params, "identifier");
            copy_param(details, &mut params, "selected_profile_uuid");
            copy_param(details, &mut params, "selected_profile_name");
            copy_param(details, &mut params, "available_profile_count");
            Some(message("yggdrasil_authenticated", params))
        }
        AuditAction::YggdrasilRefreshToken => {
            copy_param(details, &mut params, "profile_uuid");
            copy_param(details, &mut params, "profile_name");
            Some(message("yggdrasil_token_refreshed", params))
        }
        AuditAction::YggdrasilInvalidateToken => {
            copy_param(details, &mut params, "profile_uuid");
            copy_param(details, &mut params, "profile_name");
            Some(message("yggdrasil_token_invalidated", params))
        }
        AuditAction::YggdrasilSignout => {
            copy_param(details, &mut params, "profile_uuid");
            copy_param(details, &mut params, "profile_name");
            Some(message("yggdrasil_signed_out", params))
        }
        AuditAction::YggdrasilJoinServer => {
            copy_param(details, &mut params, "profile_uuid");
            copy_param(details, &mut params, "profile_name");
            copy_param(details, &mut params, "server_id_hash");
            Some(message("yggdrasil_joined_server", params))
        }
        _ => None,
    }
}

fn message(code: &str, params: BTreeMap<String, Value>) -> AuditPresentationMessage {
    AuditPresentationMessage {
        code: code.to_string(),
        params,
    }
}

fn copy_string_param(source: Option<&Value>, params: &mut BTreeMap<String, Value>, key: &str) {
    let Some(value) = source
        .and_then(|source| source.get(key))
        .and_then(Value::as_str)
    else {
        return;
    };
    params.insert(key.to_string(), Value::String(value.to_string()));
}

fn copy_param(source: &Value, params: &mut BTreeMap<String, Value>, key: &str) {
    let Some(value) = source.get(key) else {
        return;
    };
    if value.is_null() {
        return;
    }
    params.insert(key.to_string(), value.clone());
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn presentation_includes_config_key_and_value_detail() {
        let presentation = build_audit_presentation(
            AuditAction::ConfigUpdate,
            AuditEntityType::SystemConfig,
            Some(42),
            Some("audit_log_recorded_actions"),
            Some(r#"{"value":"[\"user_login\"]","visibility":"private"}"#),
        )
        .expect("presentation should be built");

        assert_eq!(presentation.summary.as_ref().unwrap().code, "config_update");
        assert_eq!(
            presentation.summary.as_ref().unwrap().params.get("key"),
            Some(&Value::String("audit_log_recorded_actions".to_string()))
        );
        assert_eq!(
            presentation.detail.as_ref().unwrap().code,
            "config_value_updated"
        );
    }

    #[test]
    fn presentation_includes_config_action_detail() {
        let presentation = build_audit_presentation(
            AuditAction::ConfigActionExecute,
            AuditEntityType::SystemConfig,
            None,
            Some("mail"),
            Some(r#"{"action":"send_test_email","target_email":"admin@example.com"}"#),
        )
        .expect("presentation should be built");

        assert_eq!(
            presentation.summary.as_ref().unwrap().code,
            "config_action_execute"
        );
        assert_eq!(
            presentation.summary.as_ref().unwrap().params.get("key"),
            Some(&Value::String("mail".to_string()))
        );
        assert_eq!(
            presentation.detail.as_ref().unwrap().code,
            "config_action_executed"
        );
        assert_eq!(
            presentation
                .detail
                .as_ref()
                .unwrap()
                .params
                .get("target_email"),
            Some(&Value::String("admin@example.com".to_string()))
        );
    }

    #[test]
    fn presentation_includes_mail_delivery_detail() {
        let presentation = build_audit_presentation(
            AuditAction::MailDeliveryFailed,
            AuditEntityType::System,
            None,
            Some("mail"),
            Some(
                r#"{"to_address":"user@example.com","template_code":"password_reset","outbox_id":7,"attempt_count":2,"error":"smtp timeout"}"#,
            ),
        )
        .expect("presentation should be built");

        assert_eq!(
            presentation.detail.as_ref().unwrap().code,
            "mail_delivery_failed"
        );
        assert_eq!(
            presentation
                .detail
                .as_ref()
                .unwrap()
                .params
                .get("to_address"),
            Some(&Value::String("user@example.com".to_string()))
        );
    }

    #[test]
    fn presentation_handles_malformed_details_with_safe_fallback_fields() {
        let presentation = build_audit_presentation(
            AuditAction::UserLogin,
            AuditEntityType::AuthSession,
            Some(7),
            Some("admin"),
            Some("not json"),
        )
        .expect("presentation should be built");

        assert_eq!(presentation.summary.as_ref().unwrap().code, "user_login");
        assert!(presentation.detail.is_none());
        assert_eq!(presentation.target.as_ref().unwrap().code, "auth_session");
    }

    #[test]
    fn presentation_uses_server_target_for_server_lifecycle_actions() {
        let presentation = build_audit_presentation(
            AuditAction::ServerStart,
            AuditEntityType::System,
            None,
            None,
            None,
        )
        .expect("presentation should be built");

        assert_eq!(presentation.summary.as_ref().unwrap().code, "server_start");
        assert_eq!(presentation.target.as_ref().unwrap().code, "server");
        assert!(presentation.detail.is_none());
    }
}

File: frontend-panel/src/lib/presentation.ts
import type {
	AuditAction,
	AuditEntityType,
	AuditLogEntry,
	AuditPresentationMessage,
	BackgroundTaskStatus,
	TaskInfo,
	TaskPresentationCode,
	TaskPresentationMessage,
} from "@/types/api";

const TASK_TITLE_LABELS = {
	runtime_task_audit_cleanup: "Audit log cleanup",
	runtime_task_auth_session_cleanup: "Auth session cleanup",
	runtime_task_background_task_dispatch: "Background task dispatch",
	runtime_task_external_auth_flow_cleanup: "External auth flow cleanup",
	runtime_task_mail_outbox_dispatch: "Mail outbox dispatch",
	runtime_task_system_health_check: "System health check",
	runtime_task_task_cleanup: "Task artifact cleanup",
	runtime_task_yggdrasil_storage_consistency_check:
		"Yggdrasil storage consistency check",
	runtime_task_yggdrasil_texture_cleanup: "Yggdrasil texture cleanup",
	runtime_task_yggdrasil_token_cleanup: "Yggdrasil token cleanup",
	status_text_failed: "Failed",
	status_text_quiet: "No changes",
	status_text_succeeded: "Succeeded",
	status_text_system_healthy: "System healthy",
	runtime_system_health_issue_detail: "System health issue",
} as const satisfies Record<TaskPresentationCode, string>;

const TASK_STATUS_LABELS = {
	pending: "Pending",
	processing: "Processing",
	retry: "Retrying",
	succeeded: "Succeeded",
	failed: "Failed",
	canceled: "Canceled",
} as const satisfies Record<BackgroundTaskStatus, string>;

const AUDIT_ACTION_LABELS = {
	admin_cleanup_tasks: "Tasks cleaned up",
	admin_create_external_auth_provider: "External auth provider created",
	admin_create_invitation: "Invitation created",
	admin_create_user: "User created",
	admin_create_user_ban: "User ban created",
	admin_create_yggdrasil_session_forward_server:
		"Yggdrasil session forwarding server created",
	admin_delete_config: "Config deleted",
	admin_delete_external_auth_provider: "External auth provider deleted",
	admin_delete_user: "User deleted",
	admin_delete_yggdrasil_session_forward_server:
		"Yggdrasil session forwarding server deleted",
	admin_disable_user: "User disabled",
	admin_revoke_invitation: "Invitation revoked",
	admin_revoke_user_ban: "User ban revoked",
	admin_revoke_user_sessions: "User sessions revoked",
	admin_test_external_auth_provider: "External auth provider tested",
	admin_update_external_auth_provider: "External auth provider updated",
	admin_update_user: "User updated",
	admin_update_user_ban: "User ban updated",
	admin_update_yggdrasil_session_forward_server:
		"Yggdrasil session forwarding server updated",
	config_action_execute: "Config action executed",
	config_delete: "Config deleted",
	config_update: "Config updated",
	external_auth_provider_create: "External auth provider created",
	external_auth_provider_delete: "External auth provider deleted",
	external_auth_provider_update: "External auth provider updated",
	mail_delivery_failed: "Email delivery failed",
	mail_send: "Email sent",
	minecraft_profile_create: "Minecraft profile created",
	minecraft_profile_delete: "Minecraft profile deleted",
	minecraft_profile_rename: "Minecraft profile renamed",
	minecraft_texture_bind: "Minecraft texture bound",
	minecraft_texture_delete: "Minecraft texture deleted",
	minecraft_texture_library_approve: "Texture library submission approved",
	minecraft_texture_library_reject: "Texture library submission rejected",
	minecraft_texture_library_submit: "Texture library submission created",
	minecraft_texture_library_unpublish: "Texture library texture unpublished",
	minecraft_texture_library_withdraw: "Texture library submission withdrawn",
	minecraft_texture_report_accept: "Texture report accepted",
	minecraft_texture_report_create: "Texture report created",
	minecraft_texture_report_reject: "Texture report rejected",
	minecraft_texture_upload: "Minecraft texture uploaded",
	server_shutdown: "Server stopped",
	server_start: "Server started",
	system_setup: "System setup",
	task_retry: "Task retry scheduled",
	user_change_password: "Password changed",
	user_confirm_email_change: "Email change confirmed",
	user_confirm_password_reset: "Password reset confirmed",
	user_confirm_registration: "Registration confirmed",
	user_external_auth_link: "External auth linked",
	user_external_auth_login: "External auth login",
	user_external_auth_unlink: "External auth unlinked",
	user_login: "User login",
	user_logout: "User logout",
	user_passkey_delete: "Passkey deleted",
	user_passkey_login: "Passkey login",
	user_passkey_register: "Passkey registered",
	user_passkey_rename: "Passkey renamed",
	user_refresh_token: "Session refreshed",
	user_register: "User registered",
	user_request_email_change: "Email change requested",
	user_request_password_reset: "Password reset requested",
	user_resend_email_change: "Email change resent",
	user_revoke_other_sessions: "Other sessions revoked",
	user_revoke_session: "Session revoked",
	user_update_profile: "Profile updated",
	yggdrasil_authenticate: "Yggdrasil login",
	yggdrasil_invalidate_token: "Yggdrasil token invalidated",
	yggdrasil_join_server: "Yggdrasil server join",
	yggdrasil_refresh_token: "Yggdrasil token refreshed",
	yggdrasil_session_forward_check: "Yggdrasil session forwarding check",
	yggdrasil_signout: "Yggdrasil signout",
} as const satisfies Record<AuditAction, string>;

const AUDIT_ENTITY_LABELS = {
	api_token: "API token",
	auth_session: "Auth session",
	external_auth_identity: "External auth identity",
	external_auth_provider: "External auth provider",
	invitation: "Invitation",
	mail: "Mail",
	minecraft_profile: "Minecraft profile",
	minecraft_texture: "Minecraft texture",
	passkey: "Passkey",
	server: "Server",
	system: "System",
	system_config: "System config",
	task: "Task",
	user: "User",
	user_ban: "User ban",
	yggdrasil_session: "Yggdrasil session",
	yggdrasil_token: "Yggdrasil token",
} as const satisfies Record<AuditEntityType | "server", string>;

const AUDIT_DETAIL_LABELS = {
	config_action_executed: "Config action executed",
	config_value_updated: "Config value updated",
	external_auth_provider_changed: "Provider settings changed",
	external_auth_provider_tested: "Provider connection tested",
	mail_delivery_failed: "Email delivery failed",
	mail_sent: "Email sent",
	minecraft_profile_renamed: "Minecraft profile renamed",
	minecraft_texture_bound: "Minecraft texture bound",
	task_retry_scheduled: "Retry queued",
	tasks_cleanup_finished: "Cleanup finished",
	user_login_identifier: "Login identifier",
	yggdrasil_session_forward_checked: "Yggdrasil session forwarding checked",
	yggdrasil_session_forward_server_changed:
		"Yggdrasil session forwarding server changed",
} as const;

type AuditDetailCode = keyof typeof AUDIT_DETAIL_LABELS;

function isRecord(value: unknown): value is Record<string, unknown> {
	return Boolean(value) && typeof value === "object" && !Array.isArray(value);
}

function messageParams(
	message:
		| AuditPresentationMessage
		| TaskPresentationMessage
		| null
		| undefined,
) {
	return isRecord(message?.params) ? message.params : {};
}

function valueText(value: unknown): string {
	if (typeof value === "string") return value.trim();
	if (typeof value === "number" || typeof value === "boolean") {
		return String(value);
	}
	if (Array.isArray(value)) {
		const values = value.flatMap((item) => {
			const text = valueText(item);
			return text ? [text] : [];
		});
		return values.length > 0 ? values.join(", ") : "";
	}
	if (isRecord(value)) {
		return JSON.stringify(value);
	}
	return "";
}

function paramText(params: Record<string, unknown>, key: string) {
	return valueText(params[key]);
}

function humanizeCode(value: string) {
	const words = value.replaceAll("-", "_").split("_").filter(Boolean);
	if (words.length === 0) return value;
	const text = words.join(" ");
	return text.charAt(0).toUpperCase() + text.slice(1);
}

function labelForCode<T extends string>(
	labels: Partial<Record<T, string>>,
	code: string,
) {
	return labels[code as T] ?? humanizeCode(code);
}

function compactJoin(
	values: Array<string | null | undefined>,
	separator = " · ",
) {
	return values
		.filter((value): value is string => Boolean(value))
		.join(separator);
}

function formatKeyValues(
	params: Record<string, unknown>,
	keys: Array<[string, string]>,
) {
	return keys
		.map(([key, label]) => {
			const value = paramText(params, key);
			return value ? `${label}: ${value}` : null;
		})
		.filter((value): value is string => Boolean(value))
		.join("; ");
}

function formatRuntimeHealthComponent(component: unknown) {
	if (!isRecord(component)) return null;
	const name = paramText(component, "name");
	const status = paramText(component, "status");
	if (!name && !status) return null;
	const message = paramText(component, "message");
	const summary = compactJoin(
		[name ? humanizeCode(name) : null, status ? humanizeCode(status) : null],
		" ",
	);
	return message ? `${summary}: ${message}` : summary;
}

function formatRuntimeHealthIssue(message: TaskPresentationMessage) {
	const params = messageParams(message);
	const components = Array.isArray(params.components)
		? params.components
				.map(formatRuntimeHealthComponent)
				.filter((component): component is string => Boolean(component))
		: [];
	if (components.length > 0) {
		return components.join("; ");
	}
	const status = paramText(params, "status");
	return status ? humanizeCode(status) : TASK_TITLE_LABELS[message.code];
}

function formatTaskPresentationMessage(
	message: TaskPresentationMessage | null | undefined,
	fallback: string,
) {
	if (!message?.code) return fallback;
	if (message.code === "runtime_system_health_issue_detail") {
		return formatRuntimeHealthIssue(message) || fallback;
	}
	if (message.code === "status_text_failed") {
		const error = paramText(messageParams(message), "error");
		return error ? `Failed: ${error}` : TASK_TITLE_LABELS[message.code];
	}
	return TASK_TITLE_LABELS[message.code] ?? fallback;
}

export function formatTaskTitle(
	task: Pick<TaskInfo, "display_name" | "presentation">,
) {
	return formatTaskPresentationMessage(
		task.presentation?.title,
		task.display_name,
	);
}

export function formatTaskStatusLabel(status: BackgroundTaskStatus) {
	return TASK_STATUS_LABELS[status] ?? humanizeCode(status);
}

export function formatTaskKind(kind: string) {
	return humanizeCode(kind);
}

export function formatTaskStatusDetail(
	task: Pick<TaskInfo, "last_error" | "presentation" | "status_text">,
) {
	const lastError = task.last_error?.trim();
	if (lastError) return lastError;

	const presentation = task.presentation?.status
		? formatTaskPresentationMessage(task.presentation.status, "")
		: "";
	if (presentation) return presentation;

	const statusText = task.status_text?.trim();
	return statusText || "-";
}

export function taskStatusBadgeVariant(status: BackgroundTaskStatus) {
	switch (status) {
		case "failed":
			return "destructive";
		case "canceled":
			return "outline";
		case "succeeded":
			return "default";
		case "pending":
		case "processing":
		case "retry":
			return "secondary";
	}
}

export function formatAuditAction(action: AuditAction | string) {
	return labelForCode<AuditAction>(AUDIT_ACTION_LABELS, String(action));
}

export function formatAuditEntityType(
	entityType: AuditEntityType | "server" | string,
) {
	return labelForCode<AuditEntityType | "server">(
		AUDIT_ENTITY_LABELS,
		String(entityType),
	);
}

function formatAuditTargetMessage(
	message: AuditPresentationMessage | null | undefined,
	fallback: string,
) {
	if (!message?.code) return fallback;
	const params = messageParams(message);
	const name = paramText(params, "name");
	const id = paramText(params, "id");
	const type = formatAuditEntityType(message.code);
	return compactJoin([
		name || (id ? null : fallback),
		type,
		id ? `#${id}` : null,
	]);
}

function formatAuditDetailMessage(
	message: AuditPresentationMessage | null | undefined,
) {
	if (!message?.code) return null;
	const params = messageParams(message);

	switch (message.code as AuditDetailCode) {
		case "config_value_updated": {
			const values = formatKeyValues(params, [
				["value", "Value"],
				["prior_visibility", "Previous visibility"],
				["visibility", "Visibility"],
			]);
			return compactJoin([AUDIT_DETAIL_LABELS.config_value_updated, values]);
		}
		case "config_action_executed": {
			const action = paramText(params, "action");
			const values = formatKeyValues(params, [
				["target_email", "Target email"],
			]);
			return compactJoin([
				action
					? `Config action: ${humanizeCode(action)}`
					: AUDIT_DETAIL_LABELS.config_action_executed,
				values,
			]);
		}
		case "external_auth_provider_changed": {
			const values = formatKeyValues(params, [
				["key", "Key"],
				["slug", "Slug"],
				["kind", "Kind"],
				["issuer_url", "Issuer"],
				["enabled", "Enabled"],
			]);
			return compactJoin([
				AUDIT_DETAIL_LABELS.external_auth_provider_changed,
				values,
			]);
		}
		case "external_auth_provider_tested": {
			const success = paramText(params, "success");
			const status = success ? (success === "true" ? "passed" : "failed") : "";
			const values = formatKeyValues(params, [
				["provider", "Provider"],
				["key", "Key"],
				["slug", "Slug"],
				["kind", "Kind"],
				["issuer_url", "Issuer"],
				["enabled", "Enabled"],
			]);
			return compactJoin([
				status
					? `Provider test ${status}`
					: AUDIT_DETAIL_LABELS.external_auth_provider_tested,
				values,
			]);
		}
		case "task_retry_scheduled": {
			const values = formatKeyValues(params, [
				["kind", "Kind"],
				["previous_attempt_count", "Previous attempts"],
			]);
			return compactJoin([AUDIT_DETAIL_LABELS.task_retry_scheduled, values]);
		}
		case "mail_sent": {
			const values = formatKeyValues(params, [
				["template_code", "Template"],
				["to_address", "To"],
				["outbox_id", "Outbox"],
			]);
			return compactJoin([AUDIT_DETAIL_LABELS.mail_sent, values]);
		}
		case "mail_delivery_failed": {
			const values = formatKeyValues(params, [
				["template_code", "Template"],
				["to_address", "To"],
				["outbox_id", "Outbox"],
				["attempt_count", "Attempts"],
				["error", "Error"],
			]);
			return compactJoin([AUDIT_DETAIL_LABELS.mail_delivery_failed, values]);
		}
		case "minecraft_texture_bound": {
			const values = formatKeyValues(params, [
				["profile_name", "Profile"],
				["profile_uuid", "Profile UUID"],
				["texture_type", "Type"],
				["texture_model", "Model"],
				["texture_hash", "Hash"],
				["width", "Width"],
				["height", "Height"],
				["file_size", "File size"],
			]);
			return compactJoin([AUDIT_DETAIL_LABELS.minecraft_texture_bound, values]);
		}
		case "minecraft_profile_renamed": {
			const values = formatKeyValues(params, [
				["old_profile_name", "Old name"],
				["new_profile_name", "New name"],
				["profile_uuid", "Profile UUID"],
				[
					"temporarily_invalidated_token_count",
					"Temporarily invalidated tokens",
				],
			]);
			return compactJoin([
				AUDIT_DETAIL_LABELS.minecraft_profile_renamed,
				values,
			]);
		}
		case "tasks_cleanup_finished": {
			const removed = paramText(params, "removed");
			const values = formatKeyValues(params, [
				["finished_before", "Before"],
				["kind", "Kind"],
				["status", "Status"],
			]);
			return compactJoin([
				removed
					? `Removed ${removed} tasks`
					: AUDIT_DETAIL_LABELS.tasks_cleanup_finished,
				values,
			]);
		}
		case "user_login_identifier": {
			const identifier = paramText(params, "identifier");
			return identifier
				? `Identifier: ${identifier}`
				: AUDIT_DETAIL_LABELS.user_login_identifier;
		}
		default:
			return labelForCode<AuditDetailCode>(AUDIT_DETAIL_LABELS, message.code);
	}
}

export function formatAuditSummary(
	entry: Pick<AuditLogEntry, "action" | "presentation">,
) {
	const message = entry.presentation?.summary;
	if (!message?.code) return formatAuditAction(entry.action);
	return formatAuditAction(message.code);
}

export function formatAuditTarget(
	entry: Pick<
		AuditLogEntry,
		"entity_id" | "entity_name" | "entity_type" | "presentation"
	>,
) {
	const fallback =
		entry.entity_name ?? formatAuditEntityType(entry.entity_type);
	if (entry.presentation?.target) {
		return formatAuditTargetMessage(entry.presentation.target, fallback);
	}
	return compactJoin([
		entry.entity_name ?? null,
		formatAuditEntityType(entry.entity_type),
		entry.entity_id == null ? null : `#${entry.entity_id}`,
	]);
}

export function formatAuditDetail(entry: Pick<AuditLogEntry, "presentation">) {
	return formatAuditDetailMessage(entry.presentation?.detail);
}

export function auditActionBadgeClass(action: AuditAction | string) {
	const value = String(action);
	if (value === "mail_delivery_failed") {
		return "border-red-200 bg-red-50 text-red-700 dark:border-red-900 dark:bg-red-950/60 dark:text-red-300";
	}
	if (value === "mail_send") {
		return "border-emerald-200 bg-emerald-50 text-emerald-700 dark:border-emerald-900 dark:bg-emerald-950/60 dark:text-emerald-300";
	}
	if (
		value.includes("delete") ||
		value.includes("disable") ||
		value.includes("revoke") ||
		value.includes("shutdown")
	) {
		return "border-red-200 bg-red-50 text-red-700 dark:border-red-900 dark:bg-red-950/60 dark:text-red-300";
	}
	if (
		value.includes("check") ||
		value.includes("login") ||
		value.includes("test")
	) {
		return "border-amber-200 bg-amber-50 text-amber-700 dark:border-amber-900 dark:bg-amber-950/60 dark:text-amber-300";
	}
	if (
		value.includes("create") ||
		value.includes("register") ||
		value.includes("setup") ||
		value.includes("update")
	) {
		return "border-sky-200 bg-sky-50 text-sky-700 dark:border-sky-900 dark:bg-sky-950/60 dark:text-sky-300";
	}
	return "border-border bg-muted/30 text-muted-foreground";
}
